internal/clients/aws.go

```diff
@@ -7,61 +7,160 @@ package clients
+			}
+		}
+		return ps, nil
+	}
+}
+
+func OptimizedTerraformSetupBuilder(ctx context.Context, c client.Client, mg resource.Managed, pc *v1beta1.ProviderConfig, ps terraform.Setup) (terraform.Setup, error) { //nolint:gocyclo
+	cfg, err := getAWSConfig(ctx, c, mg)
+	if err != nil {
+		return terraform.Setup{}, errors.Wrap(err, "cannot get AWS config")
+	}
+	ps.Configuration = map[string]any{
+		keyRegion: cfg.Region,
+	}
+
+	switch s := pc.Spec.Credentials.Source; s { //nolint:exhaustive
+	case authKeyWebIdentity:
+		if pc.Spec.Credentials.WebIdentity == nil {
+			return terraform.Setup{}, errors.New(`spec.credentials.webIdentity of ProviderConfig cannot be nil when the credential source is "WebIdentity"`)
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = pc.Spec.Credentials.WebIdentity.RoleARN
+		ps.Configuration[keyWebIdentityTokenFile] = defaultIdentityTokenFile
+	case authKeyUpbound:
+		if pc.Spec.Credentials.Upbound == nil || pc.Spec.Credentials.Upbound.WebIdentity == nil {
+			return terraform.Setup{}, errors.New(`spec.credentials.upbound.webIdentity of ProviderConfig cannot be nil when the credential source is "Upbound"`)
+		}
+		ps.Configuration[keyAssumeRoleWithWebIdentity] = pc.Spec.Credentials.Upbound.WebIdentity.RoleARN
+		ps.Configuration[keyWebIdentityTokenFile] = upboundProviderIdentityTokenFile
+	case authKeySecret:
+		data, err := resource.CommonCredentialExtractor(ctx, s, c, pc.Spec.Credentials.CommonCredentialSelectors)
+		if err != nil {
+			return terraform.Setup{}, errors.Wrap(err, "cannot get credentials")
+		}
+		cfg, err = UseProviderSecret(ctx, data, DefaultSection, cfg.Region)
+		if err != nil {
+			return terraform.Setup{}, errors.Wrap(err, errAWSConfig)
 		}
 		creds, err := cfg.Credentials.Retrieve(ctx)
 		if err != nil {
 			return terraform.Setup{}, errors.Wrap(err, "failed to retrieve aws credentials from aws config")
 		}
-		identity, err := GlobalCallerIdentityCache.GetCallerIdentity(ctx, *cfg, creds)
-		if err != nil {
-			return terraform.Setup{}, errors.Wrap(err, "cannot get the caller identity")
+		ps.Configuration = map[string]any{
+			keyRegion:          cfg.Region,
+			keyAccessKeyID:     creds.AccessKeyID,
+			keySecretAccessKey: creds.SecretAccessKey,
+			keySessionToken:    creds.SessionToken,
 		}
+	}
 
-		ps := terraform.Setup{
-			Version: version,
-			Requirement: terraform.ProviderRequirement{
-				Source:  providerSource,
-				Version: providerVersion,
-			},
-			Configuration: map[string]any{
-				keyRegion:          cfg.Region,
-				keyAccessKeyID:     creds.AccessKeyID,
-				keySecretAccessKey: creds.SecretAccessKey,
-				keySessionToken:    creds.SessionToken,
-			},
-			// Account ID is not part of provider configuration schema, so it
-			// needs to be given separately.
-			ClientMetadata: map[string]string{
-				keyAccountId: *identity.Account,
-			},
-		}
-		return ps, err
+	if len(pc.Spec.AssumeRoleChain) != 0 {
```

Comment:
We had better return an error if `len(pc.Spec.AssumeRoleChain) > 1` or there's an endpoint configuration in this path.